Sprint 4/step 1

progpspb opened the pull request.

progpspb committed:
from js to ts

progpspb committed:
edit types

gennady-bars commented on src/utils/api.ts:
```diff
@@ -8,7 +8,7 @@ export const loadIngredients = async () => {
 // создание заказов
 
-export const createOrder = async (orderDetails) => {
+export const createOrder = async (orderDetails: any) => {
```
`(orderDetails: any)` и тд. Нужно правильно типизировать данные, которые передаете в вызов функций запросов

gennady-bars commented on src/utils/api.ts:
```diff
@@ -19,11 +19,11 @@ export const createOrder = async (orderDetails) => {
 
-export const checkReponse = (res) => {
-    return res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
+export const checkReponse = (res: Response) => {
```
Отлично, что знаете про тип `Response` для ответа от сервера

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -2,28 +2,27 @@ import styles from './login.module.css';
     const { values, handleChange } = useForm({email: '', password: ''});
 
-    const handleOnSubmit = (e) => {
+    const handleOnSubmit = (e:any) => {
```
Все события обработчиков нужно **правильно** типизировать. Это легко сделать. Для `onChange` есть тип `ChangeEvent<HTMLInputElement>`, а для `onSubmit` есть `FormEvent<HTMLFormElement>` в библиотеке `react`. Для обычных функций есть тип `SyntheticEvent` для их событий. Тип `Response` для ответа от сервера, тип `KeyboardEvent` для событий клавиатуры, тип `RouteProps` для рутов

**Можно лучше**

* Есть небольшая хитрость: можно всегда проверить, какой тип у функции или события. Для этого нужно навести мышкой на атрибут `onChange`, `onSubmit`, `onClick` и тд. Там высветится типизация того, что принимают эти атрибуты.

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -3,13 +3,18 @@ import { createPortal } from 'react-dom';
     useEffect(() => {
 
-        function closeModal(event) {
+        function closeModal(event: KeyboardEvent) {
```
Отлично, что знаете про тип `KeyboardEvent` для событий клавиатуры

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -1,10 +1,11 @@
 
-import PropTypes from 'prop-types';
 import { Navigate, useLocation } from 'react-router-dom';
 import { useSelector } from 'react-redux';
```
**Можно лучше**

`import { useDispatch, useSelector } from 'react-redux';` такого больше не должно быть в проекте.  Хуки `useDispatch, useSelector` нужно типизировать **1 раз** в отдельном файле и заменить их во всем проекте на свои, а не из библиотеки `react-redux` .  Вот ссылка на документацию по их типизации https://redux.js.org/usage/usage-with-typescript#define-typed-hooks

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -28,15 +29,10 @@ const ProtectedRoute = ({ component, needAuth }) => {
     return component;
 }
 
```
**Можно лучше**

`(state: any)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что знаете про тип `Response` для ответа от сервера
* Отлично, что знаете про тип `KeyboardEvent` для событий клавиатуры

но есть некоторые недочеты:


* `(orderDetails: any)` и тд. Нужно правильно типизировать данные, которые передаете в вызов функций запросов
* Все события обработчиков нужно **правильно** типизировать. Это легко сделать. Для `onChange` есть тип `ChangeEvent<HTMLInputElement>`, а для `onSubmit` есть `FormEvent<HTMLFormElement>` в библиотеке `react`. Для обычных функций есть тип `SyntheticEvent` для их событий. Тип `Response` для ответа от сервера, тип `KeyboardEvent` для событий клавиатуры, тип `RouteProps` для рутов

**Можно лучше**

* Есть небольшая хитрость: можно всегда проверить, какой тип у функции или события. Для этого нужно навести мышкой на атрибут `onChange`, `onSubmit`, `onClick` и тд. Там высветится типизация того, что принимают эти атрибуты.

progpspb committed:
after review 1

Исправил типизацию:
1. типизировал данные, передающихся в функции запросов
2. исправлена типизация cобытий обработчиков
3. типизировал и вынес в /hooks useDispatch и useSelector
4. сделал часть из 5 спринта(типизировал actions и reducers)

gennady-bars commented on src/pages/forgot-password/forgot-password.tsx:
```diff
@@ -10,7 +11,7 @@ export default function ForgotPassword() {
     const { values, handleChange } = useForm({email: ''});
 
-    const handleOnSubmit = (e: React.SyntheticEvent) => {
+    const handleOnSubmit = (e: SyntheticEvent) => {
```
`FormEvent<HTMLFormElement>` тут нужно было использовать

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
